paderbox/utils/nested.py

```diff
@@ -400,3 +413,196 @@ def squeeze_nested(orig):
         if squeezed and all([orig[key] == orig[keys[0]] for key in keys]):
             return orig[keys[0]]
     return orig
+
+
+def _get_by_path(
+        d: Union[Mapping, Sequence],
+        path: Tuple[Any, ...],
+        broadcast: bool = False
+):
+    """ Helper function for `get_by_path` and `set_by_path`. """
+    for k in path:
+        try:
+            d = d[k]
+        except KeyError:
+            # Not sure if broadcasting makes sense for lists/tuples
+            if broadcast and not isinstance(d, dict):
+                return d
+            raise
+        except TypeError as e:
+            # Not sure if broadcasting makes sense for lists/tuples
+            if broadcast and not isinstance(d, dict):
+                return d
+            raise TypeError(
+                'Only indexable types are supported in `get_by_path` and '
+                '`set_by_path`!',
+                d, k
+            ) from e
+    return d
+
+
+def get_by_path(
+        d: Union[Mapping, Sequence],
+        path: Union[str, Tuple[Any, ...], None],
+        *,
+        broadcast: bool = False,
+        delimiter: str = '.',
+        default: Any = ...,
+) -> Any:
+    """
+    Gets a value from the nested dictionary `d` by the dotted path `path`.
+
+    Args:
+        d: The container to get the value from
+        path: Dotted path or tuple of keys to index the nested container with.
+            A `tuple` is useful if not all keys are strings. If it is a `str`,
+            keys are delimited by `delimiter`
+        broadcast: If `True`, broadcast leaves if a sub-path of `path` points to
+            a leaf in `d`. Useful for nested structures where the exact
+            structure can vary, e.g., in a database the number of samples for
+            the "observation" can be located in "num_samples" or
+            "num_samples.observation". Use with care!
+        delimiter: The delimiter for keys in path
+        default: Default value that is returned when the path is not present in
+            the nested container (and cannot be broadcasted if `broadcast=True`)
+
+    Returns:
+        Value located at `path` in `d`
+
+    Examples:
+        >>> d = {'a': 'b', 'c': {'d': {'e': 'f'}, 'g': [1, [2, 3], 4]}}
+        >>> get_by_path(d, 'a')
+        'b'
+        >>> get_by_path(d, 'c.d.e')
+        'f'
+        >>> get_by_path(d, ('c', 'g', 1, 0))
+        2
+        >>> get_by_path(d, 'a.b.c', broadcast=True)
+        'b'
+        >>> get_by_path(d, 'c.b.c', default=42)
+        42
+    """
+    if path is None:
+        return d
+    if isinstance(path, str):
+        path = path.split(delimiter)
```

Comment:
Move this code to the helper? This code is in both function and `_get_by_path` is not recursive.

Reply:
Moved the helper into `get_by_path`